Add `PaCMAP` to Seurat-Wrappers

hyhuang00 opened the pull request:
Dear Seurat-wrappers Developers,

We would like to add the `PaCMAP` algorithm to the Seurat Wrappers. This PR implements the `RunPaCMAP()` wrapper for the `PaCMAP` algorithm. The original implementation of the algorithm can be found at [the current official repository](https://github.com/YingfanWang/PaCMAP). A detailed study on PaCMAP on scRNA-seq data can be found in [our paper](https://www.nature.com/articles/s42003-022-03628-x).

Best,
Yiyang Sun, Haiyang Huang, Gaurav Rajesh Parikh, Cynthia Rudin

dcollins15 commented on R/pacmap.R:
```diff
@@ -0,0 +1,145 @@
+#' https://github.com/zhanghao-njmu/SCP/. We made modifications to ensure compatibility
+#' across multiple platforms, including Windows and macOS.
+#'
+#' @param object An object. This can be a Seurat object or a matrix-like object.
+#' @param reduction A character string specifying the reduction to be used. Default is "pca".
+#' @param dims An integer vector specifying the dimensions to be used. Default is NULL.
+#' @param features A character vector specifying the features to be used. Default is NULL.
+#' @param assay A character string specifying the assay to be used. Default is NULL.
+#' @param slot A character string specifying the slot name to be used. Default is "data".
+#' @param n_components An integer specifying the number of PaCMAP components. Default is 2.
+#' @param n.neighbors An integer specifying the number of neighbors considered in the k-Nearest Neighbor graph. Default to 10 for dataset whose sample size is smaller than 10000. For large dataset whose sample size (n) is larger than 10000, the default value is: 10 + 15 * (log10(n) - 4).
+#' @param MN_ratio A numeric value specifying the ratio of the ratio of the number of mid-near pairs to the number of neighbors. Default is 0.5.
+#' @param FP_ratio A numeric value specifying the ratio of the ratio of the number of further pairs to the number of neighbors. Default is 2.
+#' @param distance_method A character string specifying the distance metric to be used. Default is "euclidean".
+#' @param lr A numeric value specifying the learning rate of the AdaGrad optimizer. Default is 1.
+#' @param num_iters An integer specifying the number of iterations for PaCMAP optimization. Default is 450.
+#' @param apply_pca A logical value indicating whether pacmap should apply PCA to the data before constructing the k-Nearest Neighbor graph. Using PCA to preprocess the data can largely accelerate the DR process without losing too much accuracy. Notice that this option does not affect the initialization of the optimization process. Default is TRUE.
+#' @param init A character string specifying the initialization of the lower dimensional embedding. One of "pca" or "random". Default is "random".
+#' @param reduction.name A character string specifying the name of the reduction to be stored in the Seurat object. Default is "pacmap".
+#' @param reduction.key A character string specifying the prefix for the column names of the PaCMAP embeddings. Default is "PaCMAP_".
+#' @param verbose A logical value indicating whether to print verbose output. Default is TRUE.
+#' @param seed.use An integer specifying the random seed to be used. Default is 11.
+#' @param ... Additional arguments to be passed to the pacmap.PaCMAP function.
```
It's best to document parameters that are shared across all implementations (i.e., those explicitly in the method signature) in the generic function.

For example:

- Keep the description of the object parameter in the generic.
- Fully document the parameters for RunPaCMAP.default.
- Document Seurat-specific parameters in RunPaCMAP.Seurat.

dcollins15 reviewed (changes requested):
Thanks for the fantastic tool and the excellent PR! This is shaping up to be a great addition to `SeuratWrappers` 🔥 

I've got a few small tweaks for you, but generally, things are looking very clean—the main ToDos are:

1. Updating the `Version` and `Date` fields in the package's `DESCRIPTION` file to `0.4.0` and today's date respectively. 
2. Regenerating the package's man/*.Rd files — see [Roxygen2's documentation](https://roxygen2.r-lib.org/) for guidance.

Once those updates are taken care of, this should be good to go 🚀 

One optional sugggestion:

Currently, the vignette includes only one example of `RunPaCMAP` using the `features` parameter. It might be helpful for users if you add an example demonstrating its usage with the `dims` parameter as well.

---

@rsatija Just flagging that we're almost certainly going to have to update this vignette before we're able to get the `test-vignettes` check back up and running for this repo. If we'd like to take the opportunity to come up with a proper solution for resolving `python` dependencies, this would probably also benefit folks running `scVI` 🤔

hyhuang00 commented:
Hi David,

Thank you so much for your prompt review and useful advice regarding our code! Yiyang and I have made some changes accordingly. Major changes include:
- We've added a new warning message whem both `assay` and `dims` are specified.
- We separated the documentation to each function.
- We added an example in case `dims` are specified.
- We updated the `NAMESPACE` and `DESCRIPTION` via `roxygen::roxygenize()`.

Please let us know if further changes are needed!

dcollins15 reviewed (approved):
LGTM 🚀

hyhuang00 commented:
Apologies for opening a new pull request. We realized that the README update for the added method was overlooked in this PR. PR (#206) includes the missing documentation. Thank you again for your prompt review and help!
